warp/yul/ToCairoVisitor.py

```diff
@@ -1,114 +1,195 @@
 
-    def visit_expression_statement(self, node: ast.ExpressionStatement):
+    def visit_expression_statement(self, node: ast.ExpressionStatement) -> str:
         if isinstance(node.expression, ast.FunctionCall):
-            self.visit(node.expression)
+            return self.visit(node.expression)
         else:
             # see ast.ExpressionStatement docstring
             raise ValueError("What am I going to do with it? Why is it here?..")
 
-    def visit_variable_declaration(self, node: ast.VariableDeclaration):
+    def visit_variable_declaration(self, node: ast.VariableDeclaration) -> str:
+        self.preamble = False
         if node.value is None:
             decls_repr = "\n".join(
                 self.print(ast.VariableDeclaration(variables=[x], value=ast.Literal(0)))
                 for x in node.variables
             )
-            self.repr_stack.append(decls_repr)
-            return
-        vars_repr = ", ".join("local " + self.print(x) for x in node.variables)
+            self.preamble = True
+            return decls_repr
         value_repr = self.print(node.value)
         if isinstance(node.value, ast.FunctionCall):
-            if not node.variables:
-                self.repr_stack.append(value_repr)
+            function_name: str = node.value.function_name.name
+            function_args: str = ", ".join(self.print(x) for x in node.value.arguments)
+            vars_repr = self.generate_ids_typed(node.variables, function_name)
+            if function_name in YUL_BUILTINS_MAP.keys():
+                builtin_to_cairo = YUL_BUILTINS_MAP[function_name](function_args)
+                merge_imports(self.imports, builtin_to_cairo.required_imports())
+                if not node.variables:
+                    self.preamble = True
+                    return builtin_to_cairo.generated_cairo
+                else:
+                    self.preamble = True
+                    return f"""{builtin_to_cairo.preamble}
+let ({vars_repr}) = {builtin_to_cairo.function_call}
+{builtin_to_cairo.ref_copy}"""
             else:
-                self.repr_stack.append(f"let ({vars_repr}) = {value_repr}")
+                if not node.variables:
+                    self.preamble = True
+                    return value_repr
+                else:
+                    self.preamble = True
+                    return f"let ({vars_repr}) = {value_repr}"
         else:
-            self.repr_stack.append(f"{vars_repr} = {value_repr}")
+            vars_repr = self.generate_ids_typed(node.variables)
+            self.preamble = True
+            return f"{vars_repr} = {value_repr}"
 
-    def visit_block(self, node: ast.Block):
-        stmts_repr = "\n".join(self.print(x) for x in node.statements)
-        self.repr_stack.append(stmts_repr)
+    def visit_block(self, node: ast.Block) -> str:
+        try:
+            stmts_repr = "\n".join(self.print(x) for x in node.statements)
+        except TypeError:
```

Comment:
Why do we catch `TypeError` here? If something throws `TypeError` it is the underlying issue that should be solved.

In that case, I have a few suggestions:
- `visit_function_definition` instead of returning appends code to `cairo_code`. You can have it return, and just remove self.cairo_code, it's not used anywhere else
- `visit_break/continue/leave` don't return anything. For now, they can return an empty string
- something else, I don't quite know what yet

Reply:
I've tried that method but with the current structure this seems like the easiest way. Also , we only care about function definitions, because the only other top level declarations are storage variables, which we are working on getting using a different method.

Reply:
The only reason for the typeError is because `visit_functino_definition` doesn't return anything.

Reply:
Unfortunately, it masks many other kinds of type errors too. Not just when a visitor doesn't return. We've had this except block 

```python
       try:
            return getattr(self, method_name)(node, *args, **kwargs)
        except AttributeError:
            return self.common_visit(node, *args, **kwargs)
```

in yul_ast.py, and it made error discovery quite unintuitive in several cases during the short life of this project. So I had to get rid of it.

Blocks contain not only top level declarations but all possible statements of the program. So each block with `break` gets replaced by nothing.

I've applied my suggestions from the above, and I still got `None`s in the list. It might be another bug and another missing block.